### PR TITLE
IGNITE-21658 Make LowWatermark a top level component

### PR DESCRIPTION
**Pull request by AMashenkov**
https://issues.apache.org/jira/browse/IGNITE-21658
Minor LowWatermark refactoring.

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkChangedListener.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * @see LowWatermark
+ */
+@FunctionalInterface
+public interface LowWatermarkChangedListener {
+    /**
+     * Low watermark changed callback.
+     *
+     * @param ts New low watermark.
+     * @return A future, which completes when event sill be processed.
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkChangedListener.java:34]:
```suggestion
     * @return A future, which completes after the event has been processed.
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java [hunk 2]
```diff
@@ -68,15 +72,15 @@ public class LowWatermark implements ManuallyCloseable {
 
     private final VaultManager vaultManager;
 
-    private final MvGc mvGc;
+    private final List<LowWatermarkChangedListener> updateListeners = new CopyOnWriteArrayList<>();
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java:75]: LowWatermark produces a single event type, while EventProducer supports multiple types.
EventProducer requires creating a enum for Event type, an Event class for event objects, which provides a factory method for EventParameters. All this stuff is required just for passing a HybridTimestamp.
I think DRY principle is overkill here?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java [hunk 3]
```diff
@@ -131,17 +140,15 @@ public void start() {
 
             LOG.info(
                     "Low watermark has been successfully retrieved from the vault and is scheduled to be updated: {}",
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java]: Fixed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermark.java:75]: Makes sense

**Review by sashapolo — APPROVED** [review]

**Pull request merged**